### PR TITLE
Add Docker probe

### PR DESCRIPTION
**Pull request by lebauce**
This Docker probes leverages the Netns probe when a new container is created, so the topology shows the veth of the containers. It keeps a cache of the MAC addresses of all the containers to enhance the flow with the ID of the container. Note that this last part is disabled at the moment as the format of the flows has changed.

### CODE REVIEW

File: topology/probes/netns.go [hunk 1]
```diff
@@ -123,7 +123,7 @@ func NewNetNsNetLinkTopoUpdater(g *graph.Graph, n *graph.Node) *NetNsNetLinkTopo
 	}
 }
 
-func (u *NetNSProbe) onNetNsCreated(path string) {
+func (u *NetNSProbe) Probe(path string) {
```
Comment (nplanel) [inline topology/probes/netns.go]: why not use Register/Unregister words ?

File: topology/probes/netns.go [hunk 2]
```diff
@@ -167,6 +167,14 @@ func (u *NetNSProbe) onNetNsDeleted(path string) {
 	delete(u.nsnlProbes, name)
 }
 
+func (u *NetNSProbe) onNetNsCreated(path string) {
```
Comment (nplanel) [inline topology/probes/netns.go]: Why keeping the old names ?

File: topology/probes/netns.go (hunk 1 above)
Comment (lebauce) [inline topology/probes/netns.go]: Register/Unregister is indeed better. Done

File: topology/probes/netns.go (hunk 2 above)
Comment (lebauce) [inline topology/probes/netns.go]: Done

File: agent/agent.go [hunk 3]
```diff
@@ -97,19 +97,27 @@ func (a *Agent) Start() {
...
+
+	if _, err := config.GetConfig().GetSection("docker"); err == nil {
+		dockerEnhancer, err := mappings.NewDockerMapper(ns)
+		if err != nil {
+			panic(err)
+		}
+		flowEnhancers = append(flowEnhancers, dockerEnhancer)
+	}
+
+	pipeline := mappings.NewFlowMappingPipeline(flowEnhancers)
```
Comment (nplanel) [inline agent/agent.go]: why not change NewFlowMappingPipeline API by a mappings.FlowEnhancer...
so the call will be, mappings.NewFlowMappingPipeline(gfe, dockerEnhancer)

File: agent/agent.go [hunk 4]
```diff
@@ -97,19 +97,27 @@ func (a *Agent) Start() {
...
 	}
 
-	pipeline := mappings.NewFlowMappingPipeline([]mappings.FlowEnhancer{gfe})
+	ns := tprobes.NewNetNSProbe(a.Graph, root)
+	nl := tprobes.NewNetLinkProbe(a.Graph, root)
+	ovs := tprobes.NewOvsdbProbe(a.Graph, root, ovsmon)
+	flowEnhancers := []mappings.FlowEnhancer{gfe}
+
+	if _, err := config.GetConfig().GetSection("docker"); err == nil {
+		dockerEnhancer, err := mappings.NewDockerMapper(ns)
```
Comment (nplanel) [inline agent/agent.go]: a call like mappings.NewDockerMapperFromConfig() will be more consistant in our code

File: agent/agent.go (hunk 3 above)
Comment (lebauce) [inline agent/agent.go]: Unfortunately, as the docker flow enhancer is optional, it doesn't make the code shorter :-/

File: agent/agent.go (hunk 4 above)
Comment (lebauce) [inline agent/agent.go]: Indeed. Done

**Comment by nplanel** [comment]
LGTM

File: flow/mappings/docker.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	// TODO: find a place to put the container ID
+	// ethFlow := f.GetStatistics().Endpoints[flow.FlowEndpointType_ETHERNET.Value()]
+	// mapper.getContainerFromMAC(ethFlow.AB.Value)
+	// mapper.getContainerFromMAC(ethFlow.BA.Value)
+}
+
+func (mapper *DockerMapper) handleDockerEvent(event *dockerclient.Event) {
+	info, err := mapper.client.InspectContainer(event.ID)
+	if err != nil {
+		return
```
Comment (safchain) [inline flow/mappings/docker.go]: Question: does it make sense to add a log here ?

File: flow/mappings/docker.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
+/*
```
Comment (safchain) [inline flow/mappings/docker.go]: I think we should move this file to topology probe as it basically retrieves topology informations. We could add metadata information about docker on the veth node ? We already have the graph mapping in order to associate flow and intf attributes.

Comment (safchain) [inline flow/mappings/docker.go]: Maybe the neutron mapping should be moved as well

**Pull request closed without merging**